### PR TITLE
Add `observer` to `Trigger`

### LINKED ISSUE #15061: Expose Trigger::ObserverTrigger::observer
## What problem does this solve or what need does it fill?

Currently when handling a trigger i know what entity it triggered on, but i dont know what entity is listening

## What solution would you like?

The data is already in the Trigger struct and so should just be exposed.

## What alternative(s) have you considered?

Passing the entity id in a move operation by spawning separately and adding observer later.

## Additional context

Any other information you would like to add such as related previous work,
screenshots, benchmarks, etc.

````
pub fn observer(&self) -> Entity {
    self.trigger.observer
}
````

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

- Fixes  #15061

## Solution

- Added `observer` to `Trigger`, which returns the entity observing the triggered event.

## Testing

- CI passed locally.

### CODE REVIEW

**Review by lewiszlw — APPROVED** [review]

**Review by ItsDoot — APPROVED** [review]

**Comment by MiniaczQ** [comment]
I understand that we need this for sake of completion, but this isn't exactly user-sided API.
Spawning observer entities is hidden behind sugar, so a lot of people can be unaware that it is a new Entity.
People might try to manually add observers to existing entities, which is a pitfall.

I'd like the docs to mention at least some of this.

**Comment by alexparlett** [comment]
> I understand that we need this for sake of completion, but this isn't exactly user-sided API. Spawning observer entities is hidden behind sugar, so a lot of people can be unaware that it is a new Entity. People might try to manually add observers to existing entities, which is a pitfall.
> 
> I'd like the docs to mention at least some of this.

This is kind of defeating the point of reactivity, an observer wants to be able to update an entity when an event happens, the `target` is where the event is triggered, the `listener` is where the event wants to be handled, for example take progress bar with a percentage in the middle for example.

Say I have an entity model the player, the player has a resource component like stamina, that stamina is represented by a stamina bar on the top right of the screen on a different entity.

When the stamina changes on the player entity, I want the stamina bar to change too.

I can do this via a classic system, however it's a good case for observers a reactivity.

I want my observers to know what entity to update when a target entity is changed, otherwise you're only dealing in singleton components, or resources, or have to store the entity to update as another component of the target of the trigger.

An aside, i think this is even more important with the propagation changes, the target of the trigger is potentially a completely different entity to the one being watched.

**Comment by MiniaczQ** [comment]
First of all, here's the reason why you *should not* add observers onto normal entities:
https://github.com/bevyengine/bevy/issues/14204#issuecomment-2212515944
tl;dr: You can't have multiple (working) observers on one entity, that's why main API creates a new entity for each observer.

> This is kind of defeating the point of reactivity, an observer wants to be able to update an entity when an event happens, the `target` is where the event is triggered, the `listener` is where the event wants to be handled.

Adding documentation about pitfalls does not defeat the point of reactivity.
Everyone can do whatever they want in their projects, but they should be aware of APIs that are intended and APIs that are possible, but need to be handled with care.

> I want my observers to know what entity to update when a target entity is changed, otherwise you're only dealing in singleton components, or resources, or have to store the entity to update as another component of the target of the trigger.
> You will (and I have) as you say find people inserting observers directly on an existing entity to solve this problem, it doesn't make sense to not be able to react on a specific entity to a trigger happening on a specific target.

You can pass arbitrary entities inside the event data, it doesn't need to come from observer directly.

> Getting around that limitation by directly inserting observers on existing entities, aware of the deletion detection, hence the change request.

To delete observers with entities I'd suggest adding observer child entities and despawning recursively.
This isn't fully possible as of now, since only few `observe` methods return the newly created observer entity.

**Comment by alexparlett** [comment]
> Adding documentation about pitfalls does not defeat the point of reactivity.
> Everyone can do whatever they want in their projects, but they should be aware of APIs that are intended and APIs that are possible, but need to be handled with care.

Sorry you're entirely right, I didn't mean that that wasnt valuable at all. My response was more aimed at the concept og listeners only knowing about their target and that wasnt the right place for that discussion, my apologies. I'll take this discussion to a new issue.

I still think this has value, if you manually manage the creation of world scope observers, there is potentially value on adding custom components to those to help you achieve the logic you wish. To enable fetching of those components having easy access to that observer entity would be useful.

**Comment by NthTensor** [comment]
If this lets observers delete themselves (it looks like that's the case) then this is definitely needed not just a nice-to-have.

**Comment by bushrat011899** [comment]
I've updated the documentation to explain _why_ you might want to use this method (self-deletion of an observer) and included an example.

**Review by MiniaczQ — APPROVED** [review]
